File: dictionary/csharp/concurrency.cs
// Runs a task on the thread pool and returns the started task.
public static Task RunInBackground(Action task)
{
    return Task.Run(task);
}

// Awaits every task in a batch, surfacing the first failure.
public static async Task RunAllAsync(IEnumerable<Task> tasks)
{
    Task all = Task.WhenAll(tasks);
    try
    {
        await all;
    }
    catch
    {
        throw all.Exception!;
    }
}

// Runs a task with a hard timeout, cancelling on expiry.
public static async Task<T> CallWithTimeoutAsync<T>(Func<Task<T>> taskFactory, TimeSpan timeout)
{
    using var cts = new CancellationTokenSource(timeout);
    Task<T> task = taskFactory();
    Task completed = await Task.WhenAny(task, Task.Delay(Timeout.Infinite, cts.Token));
    if (completed != task)
    {
        throw new TimeoutException($"task timed out after {timeout.TotalMilliseconds}ms");
    }
    return await task;
}

// Sums a large array in parallel.
public static long ParallelSum(int[] values)
{
    return values.AsParallel().Sum(v => (long)v);
}

// Runs a task after a fixed delay.
public static async Task RunAfterDelayAsync(Action task, TimeSpan delay)
{
    await Task.Delay(delay);
    task();
}

// Thread-safe counter backed by an atomic long.
public sealed class AtomicCounter
{
    private long value;

    public long Increment()
    {
        return Interlocked.Increment(ref value);
    }

    public long Current => Interlocked.Read(ref value);
}

// Chains two async stages, transforming the result of the first.
public static async Task<string> FetchAndTransformAsync(Task<string> first)
{
    string raw = await first;
    return raw.Trim().ToUpperInvariant();
}

// Retries an operation up to attempts times with a backoff delay.
public static async Task<T> RetryAsync<T>(Func<Task<T>> action, int attempts, TimeSpan delay)
{
    Exception? last = null;
    for (int i = 0; i < attempts; i++)
    {
        try
        {
            return await action();
        }
        catch (Exception ex)
        {
            last = ex;
            await Task.Delay(TimeSpan.FromMilliseconds(delay.TotalMilliseconds * (i + 1)));
        }
    }
    throw last!;
}

// Offers into a bounded queue without blocking, failing fast when full.
public static bool OfferOrSkip<T>(System.Collections.Concurrent.BlockingCollection<T> queue, T item)
{
    return queue.TryAdd(item);
}

// Waits for all producers to signal done, with a time cap.
public static bool AwaitAll(System.Threading.CountdownEvent countdown, TimeSpan timeout)
{
    return countdown.Wait(timeout);
}

// Acquires a semaphore permit with a timeout, guarding a resource.
public static async Task<bool> TryWithPermitAsync(SemaphoreSlim semaphore, Func<Task> guarded,
    TimeSpan timeout)
{
    if (!await semaphore.WaitAsync(timeout))
    {
        return false;
    }
    try
    {
        await guarded();
        return true;
    }
    finally
    {
        semaphore.Release();
    }
}

// Applies a transform to every element in parallel.
public static List<int> ParallelMap(List<int> values, Func<int, int> transform)
{
    return values.AsParallel().Select(transform).ToList();
}

// Reads or computes a cached value atomically.
public static V CachedCompute<K, V>(
    System.Collections.Concurrent.ConcurrentDictionary<K, V> cache, K key,
    Func<K, V> loader) where K : notnull
{
    return cache.GetOrAdd(key, loader);
}

// Takes whichever task finishes first.
public static async Task<T> AwaitAnyAsync<T>(Task<T> first, Task<T> second)
{
    Task<T> completed = await Task.WhenAny(first, second);
    return await completed;
}

// Serializes writes so only one writer proceeds at a time.
public sealed class SerializedWriter
{
    private readonly object gate = new();

    public void Write(string line)
    {
        lock (gate)
        {
            Console.WriteLine($"writing: {line}");
        }
    }
}

// Runs a batch of jobs across the thread pool, collecting results.
public static async Task<List<T>> ProcessBatchAsync<T>(IEnumerable<Func<Task<T>>> jobs)
{
    var tasks = jobs.Select(job => job()).ToList();
    T[] results = await Task.WhenAll(tasks);
    return results.ToList();
}

// Doubles a value atomically, guarding against lost updates.
public sealed class AtomicDoubler
{
    private long value;

    public long DoubleAndGet()
    {
        long current = Interlocked.Read(ref value);
        while (true)
        {
            long doubled = current * 2;
            long observed = Interlocked.CompareExchange(ref value, doubled, current);
            if (observed == current)
            {
                return doubled;
            }
            current = observed;
        }
    }
}

// Runs a task with cooperative cancellation support.
public static async Task RunWithCancellationAsync(Func<CancellationToken, Task> work,
    CancellationToken token)
{
    token.ThrowIfCancellationRequested();
    await work(token);
}

// Runs two independent tasks concurrently and waits for both.
public static async Task RunConcurrentlyAsync(Func<Task> first, Func<Task> second)
{
    await Task.WhenAll(first(), second());
}

// Limits how many tasks may run at once with a fixed-permit semaphore.
public static async Task WithConcurrencyLimitAsync(int permits, IEnumerable<Func<Task>> tasks)
{
    using var semaphore = new SemaphoreSlim(permits);
    var wrapped = tasks.Select(async task =>
    {
        await semaphore.WaitAsync();
        try
        {
            await task();
        }
        finally
        {
            semaphore.Release();
        }
    });
    await Task.WhenAll(wrapped);
}

File: scripts/generate/java-data/concurrency.mjs
// Java concurrency blocks — one complete method per block.
export const concurrency = [
`// Runs a task on a fresh daemon thread.
public static Thread runInBackground(Runnable task) {
    Thread thread = new Thread(task, "worker");
    thread.setDaemon(true);
    thread.start();
    return thread;
}`,
`// Submits several tasks to a fixed pool and waits for all of them.
public static void runAll(java.util.concurrent.ExecutorService pool, List<Runnable> tasks)
        throws InterruptedException {
    List<java.util.concurrent.Future<?>> futures = new ArrayList<>();
    for (Runnable task : tasks) {
        futures.add(pool.submit(task));
    }
    for (java.util.concurrent.Future<?> future : futures) {
        try {
            future.get();
        } catch (java.util.concurrent.ExecutionException e) {
            throw new RuntimeException(e.getCause());
        }
    }
}`,
`// Runs a callable with a hard timeout, cancelling on expiry.
public static <T> T callWithTimeout(java.util.concurrent.Callable<T> task, long timeoutMillis)
        throws Exception {
    java.util.concurrent.ExecutorService pool = java.util.concurrent.Executors.newSingleThreadExecutor();
    try {
        java.util.concurrent.Future<T> future = pool.submit(task);
        return future.get(timeoutMillis, java.util.concurrent.TimeUnit.MILLISECONDS);
    } catch (java.util.concurrent.TimeoutException e) {
        throw new java.util.concurrent.TimeoutException("task timed out after " + timeoutMillis + "ms");
    } finally {
        pool.shutdownNow();
    }
}`,
`// Sums a large array in parallel using a parallel stream.
public static long parallelSum(int[] values) {
    return java.util.Arrays.stream(values).parallel().asLongStream().sum();
}`,
`// Schedules a task to run after a fixed delay.
public static java.util.concurrent.ScheduledFuture<?> runAfterDelay(
        java.util.concurrent.ScheduledExecutorService scheduler, Runnable task, long delayMillis) {
    return scheduler.schedule(task, delayMillis, java.util.concurrent.TimeUnit.MILLISECONDS);
}`,
`// Thread-safe counter backed by an atomic long.
public static class AtomicCounter {
    private final java.util.concurrent.atomic.AtomicLong value = new java.util.concurrent.atomic.AtomicLong();

    public long increment() {
        return value.incrementAndGet();
    }

    public long current() {
        return value.get();
    }
}`,
`// Chains two async stages, transforming the result of the first.
public static java.util.concurrent.CompletableFuture<String> fetchAndTransform(
        java.util.concurrent.CompletableFuture<String> first) {
    return first.thenApplyAsync(raw -> raw.trim().toUpperCase());
}`,
`// Retries an operation up to attempts times with a backoff delay.
public static <T> T retry(java.util.concurrent.Callable<T> action, int attempts, long delayMillis)
        throws Exception {
    Exception last = null;
    for (int i = 0; i < attempts; i++) {
        try {
            return action.call();
        } catch (Exception e) {
            last = e;
            Thread.sleep(delayMillis * (i + 1));
        }
    }
    throw last;
}`,
`// Offers into a bounded queue, failing fast when it is full.
public static <T> boolean offerOrSkip(java.util.concurrent.BlockingQueue<T> queue, T item) {
    return queue.offer(item);
}`,
`// Blocks until a latch reaches zero, with a time cap.
public static boolean awaitAll(java.util.concurrent.CountDownLatch latch, long timeoutMillis)
        throws InterruptedException {
    return latch.await(timeoutMillis, java.util.concurrent.TimeUnit.MILLISECONDS);
}`,
`// Acquires a permit with a timeout, guarding a shared resource.
public static boolean tryWithPermit(java.util.concurrent.Semaphore semaphore, Runnable guarded,
        long timeoutMillis) throws InterruptedException {
    if (!semaphore.tryAcquire(timeoutMillis, java.util.concurrent.TimeUnit.MILLISECONDS)) {
        return false;
    }
    try {
        guarded.run();
        return true;
    } finally {
        semaphore.release();
    }
}`,
`// Applies a transform to every element in parallel.
public static List<Integer> parallelMap(List<Integer> values, java.util.function.IntUnaryOperator fn) {
    return values.parallelStream()
            .mapToInt(fn::applyAsInt)
            .boxed()
            .collect(Collectors.toList());
}`,
`// Gracefully shuts down a pool, forcing termination on refusal.
public static void shutdownPool(java.util.concurrent.ExecutorService pool) {
    pool.shutdown();
    try {
        if (!pool.awaitTermination(5, java.util.concurrent.TimeUnit.SECONDS)) {
            pool.shutdownNow();
        }
    } catch (InterruptedException e) {
        pool.shutdownNow();
        Thread.currentThread().interrupt();
    }
}`,
`// Reads or computes a cached value atomically.
public static <K, V> V cachedCompute(
        java.util.concurrent.ConcurrentMap<K, V> cache, K key,
        java.util.function.Function<? super K, ? extends V> loader) {
    return cache.computeIfAbsent(key, loader);
}`,
`// Takes whichever future completes first.
public static <T> T awaitAny(java.util.concurrent.CompletableFuture<T> first,
        java.util.concurrent.CompletableFuture<T> second) throws Exception {
    return java.util.concurrent.CompletableFuture.anyOf(first, second).get();
}`,
`// Serializes writes with a lock so only one writer proceeds at a time.
public static class SerializedWriter {
    private final Object lock = new Object();

    public void write(String line) {
        synchronized (lock) {
            System.out.println("writing: " + line);
        }
    }
}`,
`// Runs a batch of work items across a pool, collecting results.
public static <T> List<T> processBatch(java.util.concurrent.ExecutorService pool,
        List<java.util.concurrent.Callable<T>> jobs) throws InterruptedException {
    List<java.util.concurrent.Future<T>> futures = new ArrayList<>();
    for (java.util.concurrent.Callable<T> job : jobs) {
        futures.add(pool.submit(job));
    }
    List<T> results = new ArrayList<>();
    for (java.util.concurrent.Future<T> future : futures) {
        try {
            results.add(future.get());
        } catch (java.util.concurrent.ExecutionException e) {
            throw new RuntimeException(e.getCause());
        }
    }
    return results;
}`,
`// Doubles a value atomically, guarding against lost updates.
public static class AtomicDoubler {
    private final java.util.concurrent.atomic.AtomicLong value = new java.util.concurrent.atomic.AtomicLong();

    public long doubleAndGet() {
        while (true) {
            long current = value.get();
            long doubled = current * 2;
            if (value.compareAndSet(current, doubled)) {
                return doubled;
            }
        }
    }
}`,
`// Waits for a thread to finish, ignoring interrupts.
public static void joinQuietly(Thread thread) {
    try {
        thread.join();
    } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
    }
}`,
`// Runs two independent tasks concurrently and waits for both.
public static void runConcurrently(Runnable first, Runnable second) throws InterruptedException {
    Thread a = new Thread(first);
    Thread b = new Thread(second);
    a.start();
    b.start();
    a.join();
    b.join();
}`,
`// Limits how many tasks may run at once with a fixed-permit semaphore.
public static void withConcurrencyLimit(int permits, List<Runnable> tasks)
        throws InterruptedException {
    java.util.concurrent.Semaphore semaphore = new java.util.concurrent.Semaphore(permits);
    List<Thread> threads = new ArrayList<>();
    for (Runnable task : tasks) {
        Thread thread = new Thread(() -> {
            try {
                semaphore.acquire();
                try {
                    task.run();
                } finally {
                    semaphore.release();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });
        threads.add(thread);
        thread.start();
    }
    for (Thread thread : threads) {
        thread.join();
    }
}`,
];
